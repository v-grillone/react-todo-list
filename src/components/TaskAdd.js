import React, { useState, useRef, useEffect} from 'react';
import Todo from './Todo';
import { v4 as uuidv4 } from 'uuid';

const LOCAL_STORAGE_KEY = 'todoApp.todos';

export default function TaskAdd() {
  const [todos, setTodos] = useState([
    {id:1, title: 'Gym', completed: false},
    {id:2, title: 'Homework', completed: true},
    {id:3, title: 'Homework', completed: false}
  ]);

  const taskNameRef = useRef();

  // const LOCAL_STORAGE_KEY = 'todoApp.todos';

  useEffect(() => {
    const storedTodos = JSON.parse(localStorage.getItem(LOCAL_STORAGE_KEY));
    if(storedTodos) setTodos(storedTodos);
  }, []);

  useEffect(() => {
    localStorage.setItem(LOCAL_STORAGE_KEY, JSON.stringify(todos))
  }, [todos]);

  function completeTask(id) {
    const newTodos = [...todos];
    const todo = newTodos.find(todo => todo.id === id);
    todo.completed = !todo.completed;
    setTodos(newTodos);
  }

  function clearClick() {
    const newTodos = todos.filter(todo => !todo.completed);
    setTodos(newTodos);
  }

  function handleTaskAdd(e) {
    const name = taskNameRef.current.value;
    if(name === ''){
      return
    }else{
      setTodos(prevTodos => {
        return [...prevTodos, {id: uuidv4(), title: name, completed: false}]
      });
    }
    taskNameRef.current.value = null;
  }

  return (
    <div data-testid='task-add' className='main-container flex flex-col items-center mt-10 space-y-10'>
      <div className="task-add-container space-x-4">
        <input ref={taskNameRef} placeholder='Enter a todo...' className='w-52 pl-2 border-2 border-gray-400 rounded placeholder:text-gray-300' type="text" />
        <button className='px-2 border-2 rounded border-gray-500 text-white bg-slate-400 hover:bg-gray-500' onClick={handleTaskAdd}>Add Todo</button>
      </div>
      <div className="tasks-container space-y-4">
        { todos.map((todo) => {
          return (<Todo key={todo.id} todo={todo} completeTask={completeTask}/>)
          })} 
      </div>
      <div className="clear-container">
        <button className='px-2 border-2 rounded border-red-800 text-white bg-red-300 hover:bg-red-400' onClick={clearClick}>Clear Completed</button>
      </div>
    </div>
  )
}
