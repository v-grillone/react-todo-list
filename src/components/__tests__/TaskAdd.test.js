import { render, screen , cleanup } from '@testing-library/react';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom';
import TaskAdd from '../TaskAdd';

afterEach(() => {
    cleanup();
})

test('Should render task add and updated list', () => {
    render(<TaskAdd/>);
    const taskAddElement = screen.getByTestId('task-add');
    expect(taskAddElement).toBeInTheDocument();
});

test('matches TaskAdd snapshot', () => {
    const tree2 = renderer.create(<TaskAdd/>).toJSON();
    expect(tree2).toMatchSnapshot();
})