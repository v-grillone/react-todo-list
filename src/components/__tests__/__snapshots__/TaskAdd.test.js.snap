// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches TaskAdd snapshot 1`] = `
<div
  className="main-container flex flex-col items-center mt-10 space-y-10"
  data-testid="task-add"
>
  <div
    className="task-add-container space-x-4"
  >
    <input
      className="w-52 pl-2 border-2 border-gray-400 rounded placeholder:text-gray-300"
      placeholder="Enter a todo..."
      type="text"
    />
    <button
      className="px-2 border-2 rounded border-gray-500 text-white bg-slate-400 hover:bg-gray-500"
      onClick={[Function]}
    >
      Add Todo
    </button>
  </div>
  <div
    className="tasks-container space-y-4"
  >
    <div
      className="task flex flex-row"
      data-testid="todo-1"
    >
      <h1
        className="text-2xl w-52"
      >
        GYM
      </h1>
      <button
        className="px-2 border-2 rounded border-slate-500 text-white bg-slate-400 hover:bg-slate-400"
        onClick={[Function]}
      >
        Complete
      </button>
    </div>
    <div
      className="task flex flex-row"
      data-testid="todo-2"
    >
      <strike>
        <h1
          className="text-2xl w-52"
        >
          HOMEWORK
        </h1>
      </strike>
      <button
        className="px-2 border-2 rounded border-slate-500 text-white bg-slate-400 hover:bg-slate-400"
        onClick={[Function]}
      >
        Complete
      </button>
    </div>
    <div
      className="task flex flex-row"
      data-testid="todo-3"
    >
      <h1
        className="text-2xl w-52"
      >
        HOMEWORK
      </h1>
      <button
        className="px-2 border-2 rounded border-slate-500 text-white bg-slate-400 hover:bg-slate-400"
        onClick={[Function]}
      >
        Complete
      </button>
    </div>
  </div>
  <div
    className="clear-container"
  >
    <button
      className="px-2 border-2 rounded border-red-800 text-white bg-red-300 hover:bg-red-400"
      onClick={[Function]}
    >
      Clear Completed
    </button>
  </div>
</div>
`;
